### PR TITLE
Remove Into requirement from FromHexStr

### PR DESCRIPTION
**Pull request by jeandudey**
`ToOwned` is already implemented for `str` and expresses the intent of owning the string for error reporting purpose.

IMO a lifetime would be much better to avoid resource exhaustion attacks (but it depends if the user of the API allows unstrusted input anyway) and to be `alloc`-less friendly, but lifetimes and `Error`s don't scale that well.

I'm also thinking of renaming the module to `str` as the `FromHexStr` tries to mimic `FromStr` but for hex only strings.

### CODE REVIEW

**Comment by apoelstra** [comment]
This was discussed on the original PR https://github.com/rust-bitcoin/rust-bitcoin/pull/1400#discussion_r1026777823

**Comment by apoelstra** [comment]
I'm skeptical about "resource exhaustion attacks" related to string allocations on error paths. IMO it is definitely not worth the ergonomic hit to add lifetimes to all of our error types for the sake of this error variant.

**Comment by jeandudey** [comment]
I think it's still fair to remove `Into<String>` requirement, the `MissingPrefix` error can be feature gated `#[cfg(feature = "alloc")]` in the future and have the `FromStrHex` trait be free of dependencies from `alloc`, or remove the `String` from the error, as the user anyway will know about what is the input anyway (user error handling code should give context to the error that happened, as in `Could not parse {input}, caused by: {our_error_display}`).

While resource exhaustion attacks are rare in error handling situations, it's still fairly easy to produce such an attack by passing a large enough string that when cloned triggers OOM, at least on embedded systems.

**Comment by jeandudey** [comment]
Anyway, the fewer the traits required on a function signature the more ergonomic it is to pass generic arguments on other functions, e.g.:

```rust
fn parses_something<S: AsRef<str> + Into<String>>(s: S) {
    SomeType::from_hex_str(s).unwrap()
}

// vs

fn parses_something<S: AsRef<str>>(s: S) {
    SomeType::from_hex_str(s).unwrap()
}
```

**Comment by apoelstra** [comment]
Ahh, avoiding an `alloc` dependency is a pretty cool benefit of dropping the `Into<String>`.

**Comment by apoelstra** [comment]
I'm leaning toward a "yes" on this because of this. And I agree with the suggestion that we could eventually feature-gate the error variant (though we need to be careful about not having feature-gated breaking API changes).

**Comment by apoelstra** [comment]
CI doesn't like `to_owned` for some reason but I think you could just stick with `into` actually.

**Review by apoelstra — DISMISSED** [review]
ACK 0fcc62d4a980e74bf97641e11da8cd1c06377305

**Comment by tcharding** [comment]
`ToOwned` is in prelude, to fix CI you should be able to just add `use prelude::*;`

**Comment by jeandudey** [comment]
Changed to use `prelude::*`

**Comment by Kixunil** [comment]
WTF, my response got lost.

NACK. `Into<InputString>` needs to be used instead but for that we need #1297 

For convenience we could just have a trait alias or a trait that implies `AsRef<str> + Into<InputString>`

**Comment by jeandudey** [comment]
>NACK. Into<InputString> needs to be used instead but for that we need https://github.com/rust-bitcoin/rust-bitcoin/pull/1297

I think your PR expresses my intent on my [comment](https://github.com/rust-bitcoin/rust-bitcoin/pull/1637#issuecomment-1424490346) to avoid re-allocations on error paths I guess? But for that to be useful the user would need to pass an owned `String`/`Cow<'_, str>::Owned`/`Box<str>`, passing a `&str` will always allocate, being it the common case, so, if `InputString` were to be used it'd still make sense to just `impl From<&str> for InputString` and call `s.as_ref().into()` without putting an additional constraint on the generic parameter, or remove the generic altogether and just `fn from_str_hex(s: &str)` to avoid monomorphization.  

This PR doesn't add friction with that PR IMO, this changes the current status quo until your PR is merged.

I still believe returning the input string over-complicates stuff by a lot with the only benefit being able to track which string failed the parsing. Parser code probably knows what the error is and should return the position where it happened and we can still have a something like this:

```rust
impl Error {
    /// Return a structure that implements [`Display`] to pretty-print the parsing error from
    /// the input string.
    pub fn display(&self, input: &str) -> Display {
        Display {
            input: input.to_owned(),
            ...
        }
    }
}
```

If the user wants to log the error. And doesn't force the users to our display implementation (or allocate for them if they don't want), while giving enough information about the error to properly report it (e.g.: displaying an underline where the parsing failed in an UI).

In short we should keep it simple for ergonomics, performance, security, flexibility and maintainability.

This is what `std` does: 

https://doc.rust-lang.org/std/path/struct.Path.html#method.display

**Comment by Kixunil** [comment]
> expresses my intent on my https://github.com/rust-bitcoin/rust-bitcoin/pull/1637#issuecomment-1424490346 to avoid re-allocations on error paths I guess

Kinda. We don't want lifetimes in errors, so we still want to allocate if one passes `str`. However `InputString` enables silently ignoring the error information if `alloc` is not enabled. I'd even consider a feature to skip those when allocator is enabled if it helps someone.

> if `InputString` were to be used it'd still make sense to just `impl From<&str> for InputString` and call `s.as_ref().into()` without putting an additional constraint on the generic parameter

Not really, this allocates when the user passed owned string and it shouldn't have to.

> with the only benefit being able to track which string failed

That's a fucking huge benefit. IME, number one offenders causing long and difficult debugging are incorrect and insufficient error messages. If all programs had good error messages the average debugging time would be maybe 10x less or something. Helping people print the appropriate message with the string without complication helps because people are lazy and prefer just use `?` which loses context.

Having the string inside leads to much simpler downstream code. Error paths are non-issue for maybe 99% applications. If you're one of those 1% building some real application that needs improved performance, I will happily help with getting the "do not track" feature in but not with unconditionally disabling those 99%.

**Comment by jeandudey** [comment]
>Helping people print the appropriate message with the string without complication helps because people are lazy and prefer just use ? which loses context.

With the `display` example I provided it should be a simple:

```rust
let input_string = "0xdeadbeef";
let lock_time = LockTime::from_hex_str(input_string).map_err(|e| e.display(input_string))?;
```

- Still fits on the same line of code.
- Won't allocate for the people that pass `Cow`/`String`/`Box` (`#[cfg(feature = "alloc")] fn display<I: Into<String>>(s: I)`).
- Allows just passing a `s: &str` like `FromStr::from_str` does to avoid monomorphization of the same function for 4 types (`Cow`/`String`/`Box`/`&str`).
- Doesn't disable the other 99% that want a nice error as they should have to `impl From<CannotParse> for Error` and using `?` directly will fail to compile, unless they use `Box<dyn Error>` which is already a nightmare to debug errors on.

If we assume people will be lazy to handle parsing errors, then we can safely assume that same people will be parsing untrusted input anyway and they will be passing a `&str` to the parsing function since the no allocation purpose of the `Into<InputString>` requirement is not immediately obvious, so that same people will introduce resource exhaustion attacks.

A better default is to make it obvious you are allocating with `#[cfg(feature = "alloc")] fn display<I: Into<String>>(s: I)` and if you use that it is explicitly your fault to exhaust the resources :).

**Comment by Kixunil** [comment]
In general, people will do what is simpler. So they will use `?` rather than complicated `map_err`. The right thing should be the default and reporting the error with **full** information is the right thing in 99% of cases.

> people will introduce resource exhaustion attacks.

Why do you expect that the attacker can somehow make the application construct ridiculously long string in memory and pass it into a parser but not twice as long string?

**Comment by jeandudey** [comment]
>The right thing should be the default and reporting the error with full information is the right thing in 99% of cases.

The right thing is subjective, and **full** information is too. Containing the input string doesn't give any more information, the user already has it.

I'd say the right thing is making sure the user doesn't use the rust-bitcoin in a way that can cause problems, specially for financial software. At least that is the case made for [`std::collections::HashMap`](https://doc.rust-lang.org/std/collections/struct.HashMap.html) which could use a faster hasher for better performance but could allow for HashDoS attacks, but it **allows** to use a different hasher if you don't need the DoS protections and want the extra perfomance.

>Why do you expect that the attacker can somehow make the application construct ridiculously long string in memory and pass it into a parser but not twice as long string?

If you know you can pass a somewhat not that long string to the parser and know that it causes 2 allocations for the same string all it takes is finding the right amount to do so a few times, for example, in a public API.

So the objection to a much simpler API is that the user will pass always an owned string and won't cause an extra allocation that will be done anyway since most people should be passing a `&str` since a parser shouldn't be taking ownership of data anyway for having a pretty printed error that will cost only a `map_err` or a `context` on crates like `anyhow` anyway.

**Comment by Kixunil** [comment]
> Containing the input string doesn't give any more information, the user already has it.

Absolutely not. Just try a simple experiment: make some non-trivial Rust application that accesses multiple paths in the OS and only use `?`, `io::Error` or `Box<dyn Error>` for error handling. Run the application and try to guess what got wrong once you get "File not found error". I guarantee you won't be able to know which file is missing without using strace. This is a real-life scenario that happened to me.

String inputs are even worse, because there's no `strace`. You could have a form with 20 fields and then good luck guessing which field is wrong.

> If you know you can pass a somewhat not that long string to the parser and know that it causes 2 allocations for the same string all it takes is finding the right amount to do so a few times, for example, in a public API.

It doesn't make sense to me to try to help an already-broken application to only make attack half as expensive but still possible. Factor of two in security can be rounded down to zero.

**Comment by jeandudey** [comment]
>I guarantee you won't be able to know which file is missing without using strace. This is a real-life scenario that happened to me.

```rust
    use anyhow::{Context, Result};
    let content = std::fs::read(path)
        .with_context(|| format!("Failed to read {}", path))?;
```

```
Error: Failed to read ./path/to/file

Caused by:
    No such file or directory (os error 2)
```

>It doesn't make sense to me to try to help an already-broken application to only make attack half as expensive but still possible. Factor of two in security can be rounded down to zero.

Its not about already-broken applications, its to minimize the likelihood of broken applications.

**Comment by jeandudey** [comment]
In fact, returning the input `String` is trying to poorly replicate what other crates already do and what one can do.

**Comment by Kixunil** [comment]
Yeah, importing another crate and writing extra code is required for your solution. That's my point. And as I said, IME it's really annoying. Another problem with all these crates is they don't enable localization (well, localization is another huge topic...).

**Comment by jeandudey** [comment]
>Yeah, importing another crate and writing extra code is required for your solution.

With the proposed solution on https://github.com/rust-bitcoin/rust-bitcoin/pull/1637#issuecomment-1426802900 it doesn't require an extra crate.

I just gave an example that we could give the nice error to the user while also enabling the 99% of users that will use `anyhow` and the likes.

```

**Comment by jeandudey** [comment]
>Yeah, importing another crate and writing extra code is required for your solution.

With the proposal on https://github.com/rust-bitcoin/rust-bitcoin/pull/1637#issuecomment-1426802900 it doesn't require an external crate.

And it doesn't disable the 99% of the users that will choose an error handling crate anyway. So it doesn't limit people and stuff like this can be done:

```
Error: Failed to parse lock time "0xFFAABBCCLK"

Caused by:
    Character 'L' at position 11 is not an hexadecimal digit
```

Which in my opinion doesn't get better than that for debugging.

`rust-bitcoin` doesn't have to reinvent error handling in Rust to be good at Bitcoin stuff.

**Comment by apoelstra** [comment]
I'm enjoying this debate, but to briefly get back to this PR:

* @Kixunil this PR has no behavioral changes whatsoever, it just removes a generic bound that wasn't technically necessary, even though we argued at the tme that it would always be present. Removing this makes it slightly easier to write generic code. I think it's a net win.

Then, regarding the debate at hand, @Kixunil as you observe, already users who use bare `?` will produce completely useless errors if they use std I/O anywhere. So if "users always do the simplest thing" then their crates are going to be shitty and there's nothing we can do about it. I buy @jeandudey's argument that if you want to write a serious application then you're going to need to use `anyhow` or something like it, and annotate all of your `?`s with `.with_context`s and probably lint that you don't have any bare `?`s anywhere in your application or something.

In this world, maybe it's then justifiable to make users add an extra step to annotating our errors, if (a) this lets them get "purer" errors in the case that they have their own annotation scheme, (b) it lets us avoid an `alloc` dependency to use some hex-related methods.

I don't know. I don't think the ecosystem has a good answer for this.

**Comment by Kixunil** [comment]
@jeandudey that's still too annoying to use, also the name `Display` is not generally used in this context, it's just another error type. If you think allocation should be controlled per-error-type, not globally via a feature flag, we can discuss that.

@apoelstra thinkine about it, I'd accept code that uses type ID to special-case `String`, `Box<str>` and `Cow<'_ str>`. Or maybe we could add it to `InputString` rather.

Just because I need to write a bunch of boilerplate for `io` doesn't mean I want to write even more boilerplate. There's also already an issue open at Rust to add the path information into error directly (sadly it's a huge can of worms). We can provide additional method for (a), (b) is solved by #1297 

If there's a strong distaste for this I think I could live with an opt-in feature rather than opt-out.

**Comment by jeandudey** [comment]
The name can be discussed.

I don't think we should allocate on error types and follow the same principles as in the `std` library, as there isn't a solution in the Rust community for tracking context properly. It's best to leave that option for the user I think after all.

**Comment by Kixunil** [comment]
`std` isn't some god of perfection and I don't think we're obligated to copy every design decision, just get inspired by them.

However I did notice an interesting interaction with `serde` recently where we can avoid the allocation even in the error case. I will need to think about it more.

**Review by tcharding — APPROVED** [review]
ACK 1708ddd2b677bd69ffc39c5505a6aa75db97b9ea

**Comment by jeandudey** [comment]
@Kixunil any update on this PR? I think this could still land and the `serde` approach would work without a trait requirement on the function signature.

**Comment by Kixunil** [comment]
I'm warming up to this idea. I found another interesting case: if you have a (web) form, then you don't need to repeat the input in error message because it's already visible on screen. I intend to discuss this with more Rust devs so we don't go back-and forth with APIs too much if we figure out it's still better to capture input.

**Comment by Kixunil** [comment]
I've opened a discussion at URLO: https://users.rust-lang.org/t/error-design-to-store-the-context-or-to-not/91815

**Comment by Kixunil** [comment]
I find [this response](https://users.rust-lang.org/t/error-design-to-store-the-context-or-to-not/91815/2?u=kixunil) pretty convincing. Just yesterday I was reminded how terrible error messages missing critical information are (by wasting a bunch of time). Bad error messages are a massive time waster and not worth performance speed up or theoretical DoS-resistance benefits.

I will happily support `detailed-errors` feature which turns on capturing the input string (implies `alloc`) if it means getting this in (even though I doubt the feature is actually useful).

**Comment by tcharding** [comment]
This PR has been derailed by the "should we put the source-of-error string into error variants". The PR is correct and does not change the current behaviour in any negative way. Can be merged IMO.

ACK 1708ddd2b677bd69ffc39c5505a6aa75db97b9ea

**Review by apoelstra — APPROVED** [review]
ACK 1708ddd2b677bd69ffc39c5505a6aa75db97b9ea

**Comment by Kixunil** [comment]
> But I definitely want to be able to support users who want cheap no-alloc errors.

Wouldn't it be better to have a special function for it though?

> The PR is correct and does not change the current behaviour in any negative way.

It does, it removes the possibility of avoiding the allocation which is kinda ironic since @apoelstra 
says it may be important for us.

The tradeoff here is: if you have some super weird type that's not in `std` and doesn't impl `Into<String>` you have to call the function with `&*s` or `s.as_ref()` instead of `s` but you don't allocate in error case. In other words sacrifice (uncommon) allocation to save 2-9  characters in uncommon cases. This doesn't seem worth it to me.

**Comment by jeandudey** [comment]
>It does, it removes the possibility of avoiding the allocation which is kinda ironic since @apoelstra
says it may be important for us.

It doesn't, that is assuming on your part that the error always will return the `String` and one passes a `String` to the function, which is uncommon, and a bad practice may I say because a function that parses strings shouldn't ever need ownership of the data.

The general rule for parameters is that if you're going to take ownership of something is because you absolutely need to, parsing a string doesn't need that, a `&str` is sufficient.

>Wouldn't it be better to have a special function for it though?

But why? This is a library, not a end user application that needs to show error to users. Error handling will vary a lot per application.

Most of the users will use whatever error handling crate is popular that adds context and this library, and if they don't, assuming they want to handle errors more manually then they don't need the added context of the input string anyway.

**Comment by Kixunil** [comment]
> a bad practice may I say because a function that parses strings shouldn't ever need ownership

That's misrepresentation of the correct practices. This is bad practice:

```rust
fn foo(s: &str) {
   let s = s.to_owned();
   // rest of the code that really requires owned String
}
```

This should be used instead:

```rust
fn foo(s: String) {
    // the code that requires owned value
}
```

Also this is bad practice:

```rust
fn foo(s: String) {
    println!("foo: {}", s);
}
```

Should use `impl Display`; if it needs `&str` it should require `&str`

However in our case:

* We *sometimes* do need `String`, sometimes not
* The caller *may not* need the string after the call - this is actually very common! E.g. in case of parsing command line arguments

Thus this is good practice and everything else would be bad.

Anyway, I don't understand why you're complaining about function taking ownership **if the caller chooses so**. Passing `&str` works just fine, so the callers that do need to keep using it will. I really don't get how this is better than forcing allocation to **always** happen.

> But why?

Because people might want to do both in single codebase. Features either turn it on for whole codebase or don't. A good example is making efficient serde serialization but reporting nice messages in UI.

>  use whatever error handling crate is popular that adds context and this library, and if they don't, assuming they want to handle errors more manuall

You have it reversed, calling `.context(...)?` is manual, just using `?` without `context` is automatic.

**Comment by jeandudey** [comment]
>That's misrepresentation of the correct practices. This is bad practice:

Correct practice is, if you're going to mutate the data and need to own it, take the owned value, if not, just use `&str`.

Otherwise instead of using `AsRef` the function signature would be `from_hex_str(s: String)` forcing everyone to allocate just in case there's an error.

>The caller may not need the string after the call - this is actually very common! E.g. in case of parsing command line arguments

That's not the case when for example, parsing a substring of a large piece of data that just happens to contain hex values, which IMO is more common.

>Because people might want to do both in single codebase. Features either turn it on for whole codebase or don't. A good example is making efficient serde serialization but reporting nice messages in UI.

The problem with a feature is that on final binary B I don't want dependency A that uses rust-bitcoin to enable error debugging that requires `alloc` because the platform I'm targeting doesn't have it.

But I could live with that fact if that means that it is opt-in instead forcing everyone to use `alloc`.

>You have it reversed, calling .context(...)? is manual, just using ? without context is automatic.

Chances are that if a crazy large input is passed ever to a `FromStrHex` it will just make the output error display unreadable, that's not good UX.

It can happen on the command line example, just pass some crazy values and it will show an error that doesn't make any sense.

Instead with an error handling crate it would show the error message to show **why it failed** then the input data can be attached to show what why the error happened.

On some UIs there would be a drop-down to show the attached input if it is too large to show, adding a feature for this and forcing to always display the input steals that option from users, that's not the point of a library.

Some may use `miette` to show errors on entire files and may need some information from our errors to do that.

I think the point is to make a library, not a end user application.

**Comment by Kixunil** [comment]
> Correct practice is, if you're going to mutate the data and need to own it, take the owned value, if not, just use `&str`.

Yes.



> Otherwise instead of using `AsRef` the function signature would be `from_hex_str(s: String)` forcing everyone to allocate just in case there's an error.

Indeed. Our case is none of them. We do need to own it sometimes and sometimes we don't. `AsRef<str> + Into<String>` is the most opportunistic solution.

> parsing a substring of a large piece of data that just happens to contain hex values, which IMO is more common.

It doesn't matter if it's more common, just the other case being common enough suffices. Also subsring has always type `&str` which does implement `Into<String>` so it's always viable.

> The problem with a feature is that on final binary B I don't want dependency A that uses rust-bitcoin to enable error debugging that requires `alloc` because the platform I'm targeting doesn't have it.

Such crate would do it either because it had to (wants the string) or because it's a mistake. I'm sure that in the latter case you'd find it easy to fix in a PR. In the former case you're screwed anyway.

But you're suggesting you want `alloc` to be optional. I hate to break it to you but this whole crate is already completely unusable without `alloc`. I actually want to do something about it - by separating out things like `Amount` which shouldn't require `alloc`. I recommend you start there with your contribution and then we can do the `InputString` stuff from my other PR which should make all people happy.

> and forcing to always display

Yeah, I was already thinking of having a special method that displays the message without the input string. But maybe the error messages need to be translatable which is another can of worms. For `miette` and `codespan-reporting` we need to store ranges, which I've already proposed in #1406 I actually started playing with `Amount` but didn't have time to finish.

**Comment by jeandudey** [comment]
>I hate to break it to you but this whole crate is already completely unusable without alloc.

And that isn't news to me, that is the reason of this small PR, reducing the `alloc` presence **today** instead of doing it tomorrow when Rust finally adds whatever is needed to tie errors to inputs, or forcing everyone to use the method not everyone but you prefer.

**Comment by Kixunil** [comment]
It works the other way around: until there's explicit support for no-alloc (which I was pushing for a while but needs crate split up to happen) nobody will care about this. And`InputString` solves the problem but it's kinda pointless to add it until this is no-alloc.

**Comment by jeandudey** [comment]
>until there's explicit support for no-alloc (which I was pushing for a while but needs crate split up to happen)

IMO the entire crate doesn't need to be all at once no-alloc compatible, just expose the types that are no-alloc already and just hide the others behind an `alloc` feature and gradually move code testing what works and what doesn't, can happen at the same time the crate is split up.

**Comment by Kixunil** [comment]
I've already tried something similar: to remove the requirement of having `no-std` feature. The changes were *massive* and I just gave up. Really, if you think it's easy go ahead and make a PR. :)

**Comment by stevenroose** [comment]
Tbh, the desire to print the erronous hex in the error is deviating from Rust standard practice. I don't remember where but stdlib is explicit about now needing to provide the input value into the error, that is the caller's responsibility. Integer conversion/math errors never mention integers involved, `std::path` method never contain any info about path used, ...

So we are doing something unexpected (100% such a trait requirement is unexpected for a user) just to deviate from standard practice. And tbh hex values inside errors are almost never useful, the only time I think they would be valuable is in parsing large JSON, but that's a serde issue IMO, serde should be better at indicating which value failed instead of saying "invalid hex" when one of the 5 hex fields fails. It should f.e. print the key or something, but again that's a serde issue IMO.

**Comment by apoelstra** [comment]
I don't think stdlib is "standard practice". I think stdlib's practices are widely considered to be bad, and to directly cause people to have negative impressions of Rust software.

And all serde can do, when given a string to decode, is pass to a hex parser. So there's no way it can provide better error information unless the hex parser provides it.

**Comment by stevenroose** [comment]
> I don't think stdlib is "standard practice". I think stdlib's practices are widely considered to be bad, and to directly cause people to have negative impressions of Rust software. "File not found" with zero additional information is the NullPointerException of Rust.
> 
> And all serde can do, when given a string to decode, is pass to a hex parser. So there's no way it can provide better error information unless the hex parser provides it.

I disagree with that. When you have `file::open("/foo")`, and you get "not found" that is sufficient, adding "/foo not found" is redundant. I reckon the UX problems regarding this is the widespread use of the `?` operator. Somewhere within a program when there is some config file expected to be in a place and it isn't, a program just using `?` and the trivial `From` for each sub-error type, will surface the "not found" error instead of adding metadata to it. Tools like `anyhow` help with not having to create endless error variants and still easily add context. On serde, on the example of struct parsing, serde has a context where it has the `MapAccess` thing filling in the struct, it could f.e. add the map key to the error if the value doesn't parse. But I think serde internally also is full of `?` and doesn't try to add layered errors, though that would be super helpful.

I mean I know the pain, I've had lots of large RPC responses come back with super unhelpful errors like "string not of even length" and the like. While adding the hex string into the error itself would help somewhat (you still need to first somehow log the input you're putting into serde in order to search for things in it), I don't think it merits constraining an API. In that case I'd much rather just reallocate the input and not bother the user with it.

Uff sorry for the rant. TLDR: error handling is rust is amazingly powerful, but it often takes too much effort to do correctly. Any application (not libs) should probably just use anyhow though and add context there.

**Comment by Kixunil** [comment]
> the UX problems regarding this is the widespread use of the `?` operator

Yeah, I think so. One of the best things about error handling became its curse. I actually stopped using `?` and started using `unwrap()` in all projects that don't need to be used by low-experienced users.

> I don't think it merits constraining an API.

Do you realize how miniscule the constraint is? The only types that are affected by this are exotic string types that don't impl `Into<String>` for some strange reason. And even in those cases it's trivially fixed by just calling `.as_ref()` on them (if it was impossible the type would be already unusable because of the `AsRef<str>` bound.)

**Comment by stevenroose** [comment]
> > the UX problems regarding this is the widespread use of the `?` operator
> 
> Yeah, I think so. One of the best things about error handling became its curse. I actually stopped using `?` and started using `unwrap()` in all projects that don't need to be used by low-experienced users.
> 
> > I don't think it merits constraining an API.
> 
> Do you realize how miniscule the constraint is? The only types that are affected by this are exotic string types that don't impl `Into<String>` for some strange reason. And even in those cases it's trivially fixed by just calling `.as_ref()` on them (if it was impossible the type would be already unusable because of the `AsRef<str>` bound.)

I don't know. This crate is a heavy user of newtypes as well, and this would make every stringy newtype that a user might define need to implement `Into<String>` next to `AsRef<str>` which I would find kinda annoying.

**Comment by Kixunil** [comment]
> this would make every stringy newtype that a user might define need to implement `Into<String>` next to `AsRef<str>` which I would find kinda annoying.

Not really, people can just pass in `foo.as_ref()` if they are too lazy about that. And proper libraries should implement this. (But again, this will change to `InputString`.)

**Comment by jeandudey** [comment]
Closing this one as this isn't really going anywhere and I'm not going to spend more time discussing whether it makes sense to return inputs in the error or not. Anyway this PR didn't affect #1297 as it is unmerged and still a draft.

I think this PR had value on reducing dependencies on `alloc` while proper solutions to this problem are implemented incrementally with the source code that exists today, because the supposed `Into<String>` advantage that is there today.

An alternative approach is to leave `FromHexStr` as is and introduce `FromHexStrMinimal` or any better name that doesn't allocate, `FromHexStr` would be implemented for types that `impl FromHexStrMinimal` and would return the nice (or not nice, but that's my opinion) errors with the input value.

But anyway, that just introduces complexity for just returning an error with a `String`, it's also complex like #1297, but at least doesn't force allocations on error paths even if another crate enables a potential `nice-errors` feature as users that need nice errors would use `FromHexStr`  and others would use `FromHexStrMinimal`.

I mean complex in the sense that returning an error with enough data is more simpler than adding traits and types to fix this.

My goal was to reduce dependencies on `alloc` so that an `alloc` feature is introduced, this with small PRs like this one.

But per @Kixunil [comment](https://github.com/rust-bitcoin/rust-bitcoin/pull/1637#issuecomment-1522773440) this is not the approach as the crate is unusable without `alloc` anyway and IIUC a solution to that would be a massive PR to fix that if I'm not mistaken.

Happy hacking
